common/thrift_router.h

```diff
@@ -361,31 +368,35 @@ class ThriftRouter {
 
    private:
     /*
-     * Filter hosts by role, and then sort them according to the following rules
-     *
-     * If role == ANY && !FLAGS_always_prefer_local_host, we sort the returned
-     * hosts first by
-     * (1) Prefer master to slave, then (2) Prefer local to non-local.
-     *
-     * Otherwise, we sort them by (2) only
+     * Filter and return hosts by role according to the following rules:
      *
+     * if role == ANY && !FLAGS_always_prefer_local_host, we sort the returned
+     *  hosts first by
+     *  (1) Prefer master to slave, then (2) Prefer local to non-local.
+     * for everything else, prefer local to non local.
      * If two hosts equal according to the sorting criteria, we randomly order
      * them
+     *
+     * In all of the above cases, if user has specified an az in "specific_az" parameter,
+     * it will be strictly followed and no hosts will be returned if none in that az.
      */
     auto filterByRoleAndSortByPreferenceAndShrinkTo(
         const std::vector<std::pair<const Host*, Role>>& host_info,
         const Role role,
         const unsigned rotation_counter,
         const std::string& segment,
-        const int shrink_target) {
+        const int shrink_target,
+        const std::string& specific_az) {
       std::vector<const Host*> v;
       if (role == Role::ANY && !FLAGS_always_prefer_local_host) {
         // prefer MASTER, then prefer groups with longer common prefix length
         std::vector<const Host*> v_s;
         for (const auto& hi : host_info) {
-          if (hi.second == Role::SLAVE) {
+          if (hi.second == Role::SLAVE &&
+              (specific_az == "" || hi.first->az.compare(specific_az) == 0)) {
             v_s.push_back(hi.first);
-          } else if (hi.second == Role::MASTER) {
+          } else if (hi.second == Role::MASTER
+                     && (specific_az == "" || hi.first->az.compare(specific_az) == 0)) {
```

Comment:
no need to do the check multiple places, consider:
```
for (const auto& hi : host_info) {
    if (specific_az != "" && hi.first->az.compare(specific_az) != 0) {
            continue
    }
    if (hi.second == Role::SLAVE ) {
    } else ...
}
```

Reply:
done